From 41ed355ce699369e5784614e9d2e67b558021d73 Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Fri, 6 Apr 2012 14:44:47 +0900
Subject: [PATCH 09/97] dfb-example: force UYVY pixelformat

set a capability for UYVY format between v4l2src and dfbvideosink
if -u option is specified though the commnad argument.
---
 ext/directfb/dfb-example.c |   18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/ext/directfb/dfb-example.c b/ext/directfb/dfb-example.c
index b1a92ce..03fce39 100644
--- a/ext/directfb/dfb-example.c
+++ b/ext/directfb/dfb-example.c
@@ -27,6 +27,7 @@ usage (char *cmd)
   printf ("  -y		y of sub surface rectangle\n");
   printf ("  -w		w of sub surface rectangle\n");
   printf ("  -h		h of sub surface rectangle\n");
+  printf ("  -u		specify uyvy as v4l2src output pixelformat\n");
   printf ("  -o		DirectFB or GStreamer option\n");
 }
 
@@ -42,6 +43,8 @@ main (int argc, char *argv[])
   int tmp_argc;
   char **tmp_argv;
   int i;
+  GstCaps *caps;
+  gboolean is_uyvy = FALSE;
 
   if ((argc == 2) && (strcmp (argv[1], "--help") == 0)) {
     usage (argv[0]);
@@ -56,7 +59,7 @@ main (int argc, char *argv[])
   memset (&rect, 0, sizeof (rect));
 
   opterr = 0;
-  while ((opt = getopt (argc, argv, "x:y:w:h:o:")) != -1) {
+  while ((opt = getopt (argc, argv, "x:y:w:h:uo:")) != -1) {
     switch (opt) {
       case 'x':
         rect.x = atoi (optarg);
@@ -70,6 +73,9 @@ main (int argc, char *argv[])
       case 'h':
         rect.h = atoi (optarg);
         break;
+      case 'u':
+        is_uyvy = TRUE;
+        break;
       case 'o':
         tmp_argv[tmp_argc++] = strdup (optarg);
         break;
@@ -123,8 +129,16 @@ main (int argc, char *argv[])
   /* Adding elements to the pipeline */
   gst_bin_add_many (GST_BIN (pipeline), src, sink, NULL);
 
-  if (!gst_element_link (src, sink))
+  if (!is_uyvy && !gst_element_link (src, sink))
     g_error ("Couldn't link videotestsrc and dfbvideosink");
+  else {
+    /* set a capability for UYVY */
+    caps = gst_caps_new_simple ("video/x-raw-yuv",
+        "format", GST_TYPE_FOURCC, GST_MAKE_FOURCC ('U', 'Y', 'V', 'Y'), NULL);
+    if (!gst_element_link_filtered (src, sink, caps))
+      g_error ("Couldn't link videotestsrc and dfbvideosink");
+    gst_caps_unref (caps);
+  }
 
   primary->Clear (primary, 0x00, 0x00, 0x00, 0xFF);
 
-- 
1.7.10.4

